lib/mllp/sender.ex

```diff
@@ -293,54 +275,78 @@ defmodule MLLP.Sender do
 
     payload = message |> MLLP.Envelope.wrap_message()
 
-    state.tcp.send(state.socket, payload)
-    |> case do
+    case state.tcp.send(state.socket, payload) do
       :ok ->
         reply = {:ok, :sent}
 
         telemetry(:received, %{response: reply}, state)
         {:reply, reply, state}
 
       {:error, reason} ->
-        telemetry(
-          :status,
-          %{status: :disconnected, error: reason, context: "send message failure"},
-          state
-        )
-
-        new_state = maintain_reconnect_timer(state)
-        reply = {:error, :send_message_failure}
-        {:reply, reply, new_state}
+        handle_send_error(state, reason)
     end
   end
 
-  def handle_info({:tcp_closed, _socket}, state) do
-    telemetry(:status, %{status: :disconnected, error: :tcp_closed}, state)
-    {:noreply, maintain_reconnect_timer(state)}
+  def handle_info(:connect_timeout, state) do
+    new_state =
+      case Process.info(state.socket) do
+        :undefined ->
+          telemetry(
+            :status,
+            %{status: :disconnected, error: :connect_timeout, context: "timeout message"},
+            state
+          )
+
+          attempt_connection(%State{state | pending_reconnect: nil})
+
+        _ ->
+          state
+      end
+
+    reply = {:error, :send_message_failure}
+    {:reply, reply, new_state}
   end
 
-  def handle_info({:tcp_error, _, reason}, state) do
-    telemetry(:status, %{status: :disconnected, error: :tcp_error, context: reason}, state)
-    {:stop, reason, state}
+  def terminate(reason, state) do
+    Logger.error("Sender socket terminated. Reason: #{inspect(reason)} State: %{inspect state}")
+    stop_connection(state, reason, "process terminated")
   end
 
-  def handle_info(:timeout, state) do
+  @should_reconnect_errors [:closed, :enotconn, :econnreset, :enetreset, :ehostdown]
```

Comment:
Here's the cases where we should attempt to reconnect IMO, but we should go over all possible  errors a  few time. Please refer to : https://www.erlang.org/doc/man/inet.html#type-posix

Reply:
In my testing, I could get `closed`, `enotconn` and `econnrefused` errors. Should we add `econnrefused` to list?